core/types/rollup_cost.go

```diff
@@ -116,9 +136,47 @@ func newL1CostFunc(l1Basefee, overhead, scalar *big.Int, isRegolith bool) l1Cost
 	}
 }
 
+var (
+	ecotoneDivisor = big.NewInt(1_000_000 * 16)
+	sixteen        = big.NewInt(16)
+)
+
+func newL1CostFuncEcotone(l1Basefee, l1BlobBasefee, l1BasefeeScalar, l1BlobBasefeeScalar *big.Int) l1CostFunc {
+	return func(costData RollupCostData) (fee, calldataGasUsed *big.Int) {
+		calldataGas := (costData.zeroes * params.TxDataZeroGas) + (costData.ones * params.TxDataNonZeroGasEIP2028)
+		calldataGasUsed = new(big.Int).SetUint64(calldataGas)
+
+		// Ecotone L1 cost function:
+		//
+		//   (gas/16)*(l1Basefee*16*l1BasefeeScalar + l1BlobBasefee*l1BlobBasefeeScalar)/1e6
+		//
+		// We divide "gas" by 16 to change from units of calldata gas to "estimated # of bytes when
+		// compressed".
+		//
+		// Function is actually computed as follows for better precision under integer arithmetic:
+		//
+		//   gas*(l1Basefee*16*l1BasefeeScalar + l1BlobBasefee*l1BlobBasefeeScalar)/16e6
+
+		calldataCostPerByte := new(big.Int).Set(l1Basefee)
+		calldataCostPerByte.Mul(calldataCostPerByte, sixteen).Mul(calldataCostPerByte, l1BasefeeScalar)
```

Comment:
Nit: I would prefer not to join `Mul` like this, to keep it more readable, since we're mutating the same var multiple times in the same line.
 `calldataCostPerByte = calldataCostPerByte.Mul(calldataCostPerByte, sixteen)`
and then separately 
`calldataCostPerByte = Mul(calldataCostPerByte, l1BasefeeScalar)`

Reply:
done